lib/paypal_express/api.rb

```diff
@@ -61,14 +62,18 @@ def purchase_payment(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb
 
         add_required_options(kb_payment_transaction_id, kb_payment_method_id, context, options)
 
-        properties        = merge_properties(properties, options)
-
-        # Can't use default implementation: the purchase signature is for one-off payments only
-        #super(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb_payment_method_id, amount, currency, properties, context)
-        gateway_call_proc = Proc.new do |gateway, linked_transaction, payment_source, amount_in_cents, options|
-          gateway.reference_transaction(amount_in_cents, options)
+        if options[:token] #baid
+          gateway_call_proc = Proc.new do |gateway, linked_transaction, payment_source, amount_in_cents, options|
+            gateway.reference_transaction(amount_in_cents, options)
+          end
+        else #hpp
+          gateway_call_proc = Proc.new do |gateway, linked_transaction, payment_source, amount_in_cents, options|
+            gateway.purchase(amount_in_cents, options)
```

Comment:
Maybe you can retrieve some of the state because the original request is [stored](https://github.com/killbill/killbill-paypal-express-plugin/blob/master/lib/paypal_express/private_api.rb#L29) in the DB?